src/main/webapp/datasetFieldForEditFragment.xhtml

```diff
@@ -1,32 +1,139 @@
-    xmlns:ui="http://java.sun.com/jsf/facelets"
-    xmlns:c="http://java.sun.com/jsp/jstl/core"
-    xmlns:p="http://primefaces.org/ui"
-    xmlns:pt="http://java.sun.com/jsf/passthrough"
-    xmlns:jsf="http://xmlns.jcp.org/jsf">
+        xmlns:h="http://java.sun.com/jsf/html"
+        xmlns:f="http://java.sun.com/jsf/core"
+        xmlns:ui="http://java.sun.com/jsf/facelets"
+        xmlns:c="http://java.sun.com/jsp/jstl/core"
+        xmlns:p="http://primefaces.org/ui"
+        xmlns:pt="http://java.sun.com/jsf/passthrough"
+        xmlns:jsf="http://xmlns.jcp.org/jsf">
 
     <ui:remove>
-    <!-- input text start UPDATE: UI:REMOVE applied due to duplicate ID errors, left code as reference incase test scripts complain -->
-    <span id="pre-input-#{dsf.datasetFieldType.name}" class="pre-input-tag"/></ui:remove>
-    
+        <!-- input text start UPDATE: UI:REMOVE applied due to duplicate ID errors, left code as reference incase test scripts complain -->
+        <span id="pre-input-#{dsf.datasetFieldType.name}" class="pre-input-tag"/></ui:remove>
+    <c:set var="displayCV" value="#{cvoc.size() > 0 and (dsfv.datasetField.datasetFieldType.name ==cvoc.get(dsfv.datasetField.datasetFieldType.parentDatasetFieldType.name).getKeys()[0]
+                    or dsfv.datasetField.datasetFieldType.name==cvoc.get(dsfv.datasetField.datasetFieldType.parentDatasetFieldType.name).getKeys()[1]
+                    or dsfv.datasetField.datasetFieldType.name==cvoc.get(dsfv.datasetField.datasetFieldType.parentDatasetFieldType.name).getKeys()[2]
+                    or dsfv.datasetField.datasetFieldType.name==cvoc.get(dsfv.datasetField.datasetFieldType.parentDatasetFieldType.name).getKeys()[3])}"/>
     <p:inputText value="#{dsfv.valueForEdit}" id="inputText" pt:aria-required="#{dsf.required}"
                  styleClass="form-control #{dsfv.datasetField.datasetFieldType.name == 'title' and DatasetPage.editMode == 'CREATE'  ? 'datasetfield-title' : ''}"
-                 rendered="#{!dsfv.datasetField.datasetFieldType.controlledVocabulary
+                 rendered="#{!dsfv.datasetField.datasetFieldType.controlledVocabulary and !displayCV
                              and (dsfv.datasetField.datasetFieldType.fieldType == 'TEXT'
                              or dsfv.datasetField.datasetFieldType.fieldType == 'INT'
                              or dsfv.datasetField.datasetFieldType.fieldType == 'FLOAT'
                              or dsfv.datasetField.datasetFieldType.fieldType == 'URL'
                              or dsfv.datasetField.datasetFieldType.fieldType == 'DATE'
                              or dsfv.datasetField.datasetFieldType.fieldType == 'EMAIL')}"/>
     <p:watermark for="inputText" value="#{dsfv.datasetField.datasetFieldType.localeWatermark}"></p:watermark>
+    <ui:fragment rendered="#{displayCV and dsfv.datasetField.datasetFieldType.name==cvoc.get(dsfv.datasetField.datasetFieldType.parentDatasetFieldType.name).getKeys()[0]}">
```

Comment:
@poikilotherm, can you please give us a good example? It looks like a cosmetic thing.

Reply:
Sure. Instead of retrieving from the `Map<String,CVoC>`, we could just do a lookup via `DatasetPage.getKeyPart(dsfv, part)`.

Of course the method name should be much more descriptive. that "key" thing is a bit unclear, also the numeric index is a bit unclear why this is necessary and what's the intention behind. The POJO definition is also only showing it's coming from the "vocab-codes".

The method itself in the backing bean (in view scope) would ask the `DatasetFieldServiceBean` for the configuration POJO, which is looked up inside (and initialized once for caching, recreated by an updated setting via API/...). 

We might need to change this stateless bean to a singleton when using as a facade. See also https://www.adam-bien.com/roller/abien/entry/singleton_the_perfect_cache_facade

The comparison could also happen inside a method of the backing bean, but I'd call that bike shedding.